src/rust/src/wallet.rs

```diff
@@ -135,33 +155,113 @@ impl Wallet {
         Wallet {
             key_store: KeyStore::empty(),
             wallet_tx_notes: HashMap::new(),
+            nullifiers: BTreeMap::new(),
             witness_tree: BridgeTree::new(MAX_CHECKPOINTS),
+            last_checkpoint: None,
             last_observed: None,
             locked_notes: HashSet::new(),
             spent_notes: HashSet::new(),
         }
     }
 
-    pub fn checkpoint_witness_tree(&mut self) {
+    /// Checkpoints the note commitment tree. This returns `false` and leaves the note
+    /// commitment tree unmodified if the block height does not immediately succeed
+    /// the last checkpointed block height (unless the note commitment tree is empty,
+    /// in which case it unconditionally succeeds). This must be called exactly once
+    /// per block.
+    pub fn checkpoint(&mut self, block_height: BlockHeight) -> bool {
+        // checkpoints must be in order of sequential block height and every
+        // block must be checkpointed
+        if self
+            .last_checkpoint
+            .iter()
+            .any(|last_height| block_height != *last_height + 1)
+        {
+            return false;
+        }
+
         self.witness_tree.checkpoint();
+        self.last_checkpoint = Some(block_height);
+        true
     }
 
-    pub fn rewind_witness_tree(&mut self) -> bool {
-        self.witness_tree.rewind()
+    /// Returns whether or not a checkpoint has been created. If no checkpoint exists,
+    /// the wallet has not yet observed any blocks.
+    pub fn is_checkpointed(&self) -> bool {
+        self.last_checkpoint.is_some()
+    }
+
+    /// Rewinds the note commitment tree to the last checkpoint. If the rewind operation
+    /// succeeds, this also decrements the checkpoint height.
+    pub fn rewind(&mut self, to_height: BlockHeight) -> bool {
+        if let Some(checkpoint_height) = self.last_checkpoint.as_mut() {
+            // if the rewind length exceeds the number of checkpoints we have stored,
+            // return failure.
+            let mut blocks_to_rewind: i64 =
+                <i64>::from(to_height) - <i64>::from(*checkpoint_height);
+            if blocks_to_rewind > self.witness_tree.checkpoints().len().try_into().unwrap() {
+                return false;
+            }
+
+            // when we rewind, we need to remove the spentness marks for the notes we're tracking
+            while blocks_to_rewind > 0 {
+                if !self.witness_tree.rewind() {
+                    return false;
+                }
+
+                *checkpoint_height = *checkpoint_height - 1;
+                for (txid, n) in self.wallet_tx_notes.iter() {
+                    if n.tx_height > *checkpoint_height {
+                        self.spent_notes.retain(|o| o.txid != *txid)
+                    }
+                }
```

Comment:
This iterates multiple times internally over `self.spent_notes`, as well as iterating over `self.wallet_tx_notes`. It would be more efficient to invert this:
```suggestion
                self.spent_notes.retain(|o| {
                    self.wallet_tx_notes[o.txid].tx_height <= *checkpoint_height
                });
```
This is single-pass over `self.spent_notes`, and `HashMap` lookups are `O(1)~`.